spatch/__init__.py

```diff
@@ -0,0 +1,212 @@
+import inspect
+import functools
+import importlib
+import textwrap
+
+
+def get_identifier(obj):
+    """Helper to get any objects identifier.  Is there an exiting short-hand?
+    """
+    return f"{obj.__module__}:{obj.__qualname__}"
+
+
+def from_identifier(ident):
+    module, qualname = ident.split(":")
+    obj = importlib.import_module(module)
+    for name in qualname.split("."):
+        obj = getattr(obj, name)
+
+    return obj
+
+
+class WillNotHandle:
+    """Class to return when an implementation does not want to handle
+    args/kwargs.
+    """
+    def __init__(self, info="<unknown reason>"):
+        self.info = info
+
+
+class Backend: 
+    @classmethod
+    def from_info_dict(cls, info):
+        return cls.from_mapping_and_types(info["name"], info["types"], info["symbol_mapping"])
+
+    @classmethod
+    def from_mapping_and_types(cls, name, types, symbol_mapping):
+        """
+        Create a new backend.
+        """
+        self = cls()
+        self.name = name
+        self.type_names = types
+        self.symbol_mapping = symbol_mapping
+        return self
+
+    def match_types(self, types):
+        """See if this backend matches the types, we do this by name.
+
+        Of course, we could use more complicated ways in the future.
+        E.g. one thing is that we can have to kind of types:
+        1. Types that must match (at least once).
+        2. Types that are understood (we do not break for them).
+
+        Returns
+        -------
+        matches : boolean
+            Whether or not the types matches.
+        unknown_types : sequence of types
+            A sequence of types the backend did not match/know.
+            This may be a way to e.g. deal with scalars, that we assume
+            all backends can convert, but creating an extensive list may
+            not be desireable?
+        """
+        matches = False
+        unknown_types = []
+        for t in types:
+            ident = get_identifier(t)
+
+            if ident in self.type_names:
```

Comment:
So, this uses the string types, and @eriknw correctly pointed out (independent of this PR), that projects may sometimes be sloppy about which `__module__` they define types.
I am not sure this is a problem, but we could work around it, with something like:
```
module, name = self.type_names[0]  # all of them of course
if module in sys.modules:
    t = getattr(module, name, None)  # ignore error?
    ...
```
I.e. I think rather than matching only by strings, we could check whether the type may be defined, by testing if the module exists.